diffrax/_solver/quicsort.py

```diff
@@ -0,0 +1,250 @@
+        a_third = jnp.stack(
+            [c4 / 29160, -c3 / 1944, c2 / 162, -c / 18, one / 3, zero], axis=-1
+        )
+        a_third = jnp.expand_dims(a_third, axis=c.ndim)
+        a_div_h = jnp.stack(
+            [-c5 / 720, c4 / 120, -c3 / 24, c2 / 6, -c / 2, one], axis=-1
+        )
+        a_div_h = jnp.expand_dims(a_div_h, axis=c.ndim)
+        assert a_third.shape == a_div_h.shape == c.shape + (1, 6)
+
+        return _QUICSORTCoeffs(
+            beta_lr1=beta_lr1,
+            a_lr1=a_lr1,
+            b_lr1=b_lr1,
+            a_third=a_third,
+            a_div_h=a_div_h,
+        )
+
+    def _compute_step(
+        self,
+        h: RealScalarLike,
+        levy: AbstractSpaceTimeTimeLevyArea,
+        x0: UnderdampedLangevinX,
+        v0: UnderdampedLangevinX,
+        uld_args: ULDArgs,
+        coeffs: _QUICSORTCoeffs,
+        rho: UnderdampedLangevinX,
+        prev_f: UnderdampedLangevinX,
+    ) -> tuple[UnderdampedLangevinX, UnderdampedLangevinX, UnderdampedLangevinX, None]:
+        dtypes = jtu.tree_map(jnp.result_type, x0)
+        w: UnderdampedLangevinX = jtu.tree_map(jnp.asarray, levy.W, dtypes)
+        hh: UnderdampedLangevinX = jtu.tree_map(jnp.asarray, levy.H, dtypes)
+        kk: UnderdampedLangevinX = jtu.tree_map(jnp.asarray, levy.K, dtypes)
+
+        gamma, u, f = uld_args
+
+        def _extract_coeffs(coeff, index):
+            return jtu.tree_map(lambda arr: arr[..., index], coeff)
+
+        beta_l = _extract_coeffs(coeffs.beta_lr1, 0)
+        beta_r = _extract_coeffs(coeffs.beta_lr1, 1)
+        beta_1 = _extract_coeffs(coeffs.beta_lr1, 2)
+        a_l = _extract_coeffs(coeffs.a_lr1, 0)
+        a_r = _extract_coeffs(coeffs.a_lr1, 1)
+        a_1 = _extract_coeffs(coeffs.a_lr1, 2)
+        b_l = _extract_coeffs(coeffs.b_lr1, 0)
+        b_r = _extract_coeffs(coeffs.b_lr1, 1)
+        b_1 = _extract_coeffs(coeffs.b_lr1, 2)
+        a_third = _extract_coeffs(coeffs.a_third, 0)
+        a_div_h = _extract_coeffs(coeffs.a_div_h, 0)
+
+        rho_w_k = (rho**ω * (w**ω - 12 * kk**ω)).ω
+        uh = (u**ω * h).ω
+        v_tilde = (v0**ω + rho**ω * (hh**ω + 6 * kk**ω)).ω
+
+        x1 = (x0**ω + a_l**ω * v_tilde**ω + b_l**ω * rho_w_k**ω).ω
+        f1uh = (f(x1) ** ω * uh**ω).ω
+
+        x2 = (
+            x0**ω + a_r**ω * v_tilde**ω + b_r**ω * rho_w_k**ω - a_third**ω * f1uh**ω
+        ).ω
+        f2uh = (f(x2) ** ω * uh**ω).ω
+
+        x_out = (
+            x0**ω
+            + a_1**ω * v_tilde**ω
+            + b_1**ω * rho_w_k**ω
+            - 0.5 * (a_r**ω * f1uh**ω + a_l**ω * f2uh**ω)
+        ).ω
+
+        v_out_tilde = (
+            beta_1**ω * v_tilde**ω
+            - 0.5 * (beta_r**ω * f1uh**ω + beta_l**ω * f2uh**ω)
+            + a_div_h**ω * rho_w_k**ω
+        ).ω
+        v_out = (v_out_tilde**ω - rho**ω * (hh**ω - 6 * kk**ω)).ω
+
+        # this method is not FSAL, but for compatibility with the base class we set
+        f_fsal = prev_f
```

Comment:
This seems dodgy. If I understand correctly then this gets piped all the way back around to `prev_f` in the entrypoint to this function, where it is unused, and is therefore technically fine, but even so...

I'd suggest setting `f_fsal = None` here and tweaking the shape-checking code to allow for that. Also, `del prev_f` at the start of this function as an assertion that it is not used.

Reply:
Yeah, that's fair.

I did have to add an `_is_fsal` class var, so that I set `solver_state.prev_f` to the right value in `AbstractFosterLangevinSRK.init` (otherwise I get `body_fun must have the same input and output structure`).

Reply:
I think that's fine!